### PR TITLE
fully_customizable_filename_format

### PR DESCRIPTION
**Pull request by npes87184**
!n=bookname, !a=author !t=yyyy-MM-dd, !y=year, !m=month and !d=day

And I use regex to validate a filename (can't contain *&%/).

### CODE REVIEW

**Comment by sh1r0** [comment]
Hi @npes87184, I'll review this pr later. Thanks for your efforts.

File: app/src/main/java/com/sh1r0/noveldroid/SettingsFragment.java [hunk 1]
```diff
@@ -81,10 +86,42 @@ public void onCreate(Bundle paramBundle) {
...
+			public boolean onPreferenceClick(Preference preference) {
+				LayoutInflater factory = LayoutInflater.from(getActivity());
+				final View dialog_view = factory.inflate(R.layout.naming_rule_dialog,null);
+				final EditText editText;
+				editText =  (EditText) dialog_view.findViewById(R.id.editText);
+				editText.setText(prefs.getString(KEY_NAMING_RULE, "!n"));
+				AlertDialog.Builder editDialog = new AlertDialog.Builder(getActivity());
+				editDialog.setTitle(getResources().getString(R.string.filename_rule_dialog_title));
+				editDialog.setView(dialog_view);
+				editDialog.setPositiveButton("OK", new DialogInterface.OnClickListener() {
```
Comment (sh1r0) [inline app/src/main/java/com/sh1r0/noveldroid/SettingsFragment.java]: Try not to use hardcoded string for UI components.

File: app/src/main/res/layout/naming_rule_dialog.xml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:orientation="vertical" android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+    <TextView
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:text="@string/name_rule"
+        android:id="@+id/textView"
```
Comment (sh1r0) [inline app/src/main/res/layout/naming_rule_dialog.xml]: Please follow the naming rule for android:id in other resource files.

File: app/src/main/res/layout/naming_rule_dialog.xml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+        android:layout_gravity="center_horizontal" />
+
+    <EditText
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:id="@+id/editText"
```
Comment (sh1r0) [inline app/src/main/res/layout/naming_rule_dialog.xml]: Please follow the naming rule for android:id in other resource files.

**Comment by npes87184** [comment]
I will edit it after dinner. 

Best

File: app/src/main/java/com/sh1r0/noveldroid/SettingsFragment.java [hunk 4]
```diff
@@ -81,10 +86,42 @@ public void onCreate(Bundle paramBundle) {
...
 		encoding = findPreference(KEY_ENCODING);
 		namingRule = findPreference(KEY_NAMING_RULE);
+		namingRule.setOnPreferenceClickListener(new OnPreferenceClickListener() {
+			@Override
+			public boolean onPreferenceClick(Preference preference) {
+				LayoutInflater factory = LayoutInflater.from(getActivity());
+				final View dialog_view = factory.inflate(R.layout.naming_rule_dialog,null);
+				final EditText editText;
+				editText =  (EditText) dialog_view.findViewById(R.id.editText);
+				editText.setText(prefs.getString(KEY_NAMING_RULE, "!n"));
```
Comment (sh1r0) [inline app/src/main/java/com/sh1r0/noveldroid/SettingsFragment.java]: Having a constant variable for this default value should provide more flexibility.

**Comment by sh1r0** [comment]
Hey @npes87184, great work. But IMHO, using `!` as the leading prefix is not good though, I would recommend you to use prohibited characters for filenames, e.g., `/`.
Thanks.

**Comment by npes87184** [comment]
Hey @sh1r0,
If use prohibited characters for filenames as the leading prefix. It is difficult to validate the filename.

For example, I should check if it contains special meaning symbol or not first.
After that, I can check whether user-defined filename string is legal or not.

**Comment by sh1r0** [comment]
@npes87184, you could try to use `File` to check the validity of filename, and I think thast would be more clear.

**Comment by npes87184** [comment]
Hi, @sh1r0, use File to check the validity of filename means this?

```
public boolean isValidFileName(final String aFileName) {
    final File aFile = new File(aFileName);
    boolean isValid = true;
    try {
        if (aFile.createNewFile()) {
            aFile.delete();
        }
    } catch (IOException e) {
        isValid = false;
    }
    return isValid;
}
```

And in onCreate 

```
novelName = "書名";
authorName = "作者名";
strDate = "2016-8-1";

...

String userDefinedName = editText.getText().toString();
userDefinedName = userDefinedName.replace("/n", novelName);
userDefinedName = userDefinedName.replace("/a", authorName);
userDefinedName = userDefinedName.replace("/t", strDate);
userDefinedName = userDefinedName.replace("/y", yyyyDate);
userDefinedName = userDefinedName.replace("/m", mmDate);
userDefinedName = userDefinedName.replace("/d", ddDate);

userDefinedName = userDefinedName + ".txt";

if(isValidFileName(userDefinedName)) {

} else {
    // ilegal ...
}
```

I know this method looks like very stupid. But, I don't have good idea about it. :(

File: app/src/main/java/com/sh1r0/noveldroid/SettingsFragment.java [hunk 5]
```diff
@@ -157,8 +195,7 @@ public void onResume() {
 		super.onResume();
 		getActivity().setTitle(R.string.settings);
 		encoding.setSummary(prefs.getString(KEY_ENCODING, "UTF-8"));
-		namingRulePref = Integer.parseInt(prefs.getString(KEY_NAMING_RULE, "0"));
-		namingRule.setSummary(namingRuleList[namingRulePref]);
+		namingRule.setSummary(prefs.getString(KEY_NAMING_RULE, DEFAULT_NAMING_RULE));
```
Comment (sh1r0) [inline app/src/main/java/com/sh1r0/noveldroid/SettingsFragment.java]: The summary here could be the preview result of the input pattern, for example, `/a_/n => author_name`, where `/` is the prefix pattern.

File: app/src/main/java/com/sh1r0/noveldroid/MainActivity.java [hunk 6]
```diff
@@ -235,7 +235,7 @@ public void run() {
 							downDirPath = prefs.getString("down_dir", NovelUtils.APP_DIR);
 							mHandler.sendEmptyMessage(PREPARING);
 							filename = novelDownloader.process(downDirPath,
-									Integer.parseInt(prefs.getString("naming_rule", "0")),
+									prefs.getString("naming_rule", "!n"),
```
Comment (sh1r0) [inline app/src/main/java/com/sh1r0/noveldroid/MainActivity.java]: Use the constant?

**Comment by sh1r0** [comment]
IMO, the following segment would be better, as there is no need to "create" a file.

``` java
public static boolean isFilenameValid(String file) {
    File f = new File(file);
    try {
        return f.getCanonicalFile().getName().equals(file);
    } catch (IOException e) {
        return false;
    }
}
```

And you could check the validity in the click event listener of `editDialog` by using the preview format (instead of the real filename) I mentioned in the comment above for a early check.

**Comment by npes87184** [comment]
Hi @sh1r0, I have finished it!

**Comment by npes87184** [comment]
Hi, @sh1r0 
Any concerns?

File: app/src/main/java/com/sh1r0/noveldroid/SettingsFragment.java [hunk 7]
```diff
@@ -107,7 +133,7 @@ public boolean onPreferenceClick(Preference preference) {
 					public void onClick(DialogInterface arg0, int arg1) {
 						String userDefinedName = editText.getText().toString();
 						String previewString = toPreviewFileNameString(userDefinedName);
-						if( isFilenameValid(previewString) ) {
+						if( isFilenameValid(previewString) && !previewString.equals("") ) {
```
Comment (sh1r0) [inline app/src/main/java/com/sh1r0/noveldroid/SettingsFragment.java]: Please follow the coding style, i.e, whitespace, etc.

File: app/src/main/java/com/sh1r0/noveldroid/SettingsFragment.java [hunk 8]
```diff
@@ -97,7 +99,31 @@ public boolean onPreferenceClick(Preference preference) {
...
 				final View dialog_view = factory.inflate(R.layout.naming_rule_dialog,null);
 				final EditText editText;
+				final TextView previewTextView;
+				previewTextView = (TextView) dialog_view.findViewById(R.id.tv_preivew_string);
 				editText =  (EditText) dialog_view.findViewById(R.id.et_naming_rule);
+				editText.addTextChangedListener(new TextWatcher() {
+
+					@Override
+					public void afterTextChanged(Editable s) {
+						// TODO Auto-generated method stub
```
Comment (sh1r0) [inline app/src/main/java/com/sh1r0/noveldroid/SettingsFragment.java]: Please remove untended empty lines and comments.

**Comment by sh1r0** [comment]
Hi @npes87184, great job! Please rebase the commits and squash them, I'll merge this PR after that.
Thanks.

**Comment by npes87184** [comment]
Hi, @sh1r0 

Done!

**Pull request merged**